src/syntax/Stratification.jl

```diff
@@ -0,0 +1,295 @@
+    @match mapping_pair begin
+
+
+        :($s => $t <= $a) => return ([DSLArgument(s,t)], [DSLArgument(a,t)])
+        :($s => $t <= $a, $(atail...)) => ([DSLArgument(s,t)], [DSLArgument(a,t) ; [DSLArgument(as,t) for as in atail] ])#return (Dict(unwrap_key_expression(s, t)), push!(Dict(unwrap_key_expression(as, t) for as in atail), unwrap_key_expression(a, t)))
+        :($(shead...), $s => $t <= $a) => ([[DSLArgument(ss, t) for ss in shead] ; DSLArgument(s, t)], [DSLArgument(a, t)])#return (push!(Dict(unwrap_key_expression(ss, t) for ss in shead), unwrap_key_expression(s, t)), Dict(unwrap_key_expression(a, t)))
+
+        if mapping_pair.head == :tuple end => begin
+            middle_index = findfirst(x -> typeof(x) == Expr && length(x.args) == 3, mapping_pair.args) # still isn't specific enough
+            if isnothing(middle_index)
+                error("Malformed line $mapping_pair, could not find center.")
+            end
+            @match mapping_pair.args[middle_index] begin
+                :($stail => $t <= $ahead) => begin
+                    sdict = [[DSLArgument(ss, t) for ss in mapping_pair.args[1:middle_index-1]] ; DSLArgument(stail, t)]
+                    adict = [DSLArgument(ahead, t) ; [DSLArgument(as, t) for as in mapping_pair.args[middle_index+1:end]]]
+                    return (sdict, adict)
+                end
+                _ => "Unknown format found for match; middle three values formatted incorrectly."
+            end
+        end
+        _ => error("Unknown line format found in stratification notation.") 
+    end
+end
+
+
+
+
+function read_stratification_line_and_update_dictionaries!(line::Expr, strata_names::Dict{Symbol, Int}, type_names::Dict{Symbol, Int}, aggregate_names::Dict{Symbol, Int}, strata_mappings::Dict{Int, Int}, aggregate_mappings::Dict{Int, Int})
+    current_strata_symbol_dict, current_aggregate_symbol_dict = interpret_stratification_notation(line)
+
+    current_strata_dict = substitute_symbols(strata_names, type_names, current_strata_symbol_dict ; use_flags=true)
+    current_aggregate_dict = substitute_symbols(aggregate_names, type_names, current_aggregate_symbol_dict ; use_flags=true)
+    
+    if STRICT_MATCHES
+        @assert (all(x -> x ∉ keys(strata_mappings), keys(current_strata_dict))) "Attempt to overwrite a mapping in strata!"
+        # check that we're not overwriting a value which has already been assigned
+        merge!(strata_mappings, current_strata_dict) # accumulate dictionary keys
+
+
+        @assert (all(x -> x ∉ keys(aggregate_mappings), keys(current_aggregate_dict))) "Attempt to overwrite a mapping in aggregate!"
+        merge!(aggregate_mappings, current_aggregate_dict)
+
+    else 
+        mergewith!((x, y) -> x, strata_mappings, current_strata_dict) # alternatively, can use: only ∘ first
+        mergewith!((x, y) -> x, aggregate_mappings, current_aggregate_dict)
+    end
+
+end
+
+
+
+
+"""
+    @stratify (strata, type, aggregate) begin ... end
```

Comment:
Yeah that's the result of a typo.  Originally, this was done using a macro called @stratify, but since I needed the definitions of the stockflow, I turned it into a function.

Similarly, composition was originally a macro @compose.

Reply:
Is there anything stopping these from being macros? I think we can still have parameters (definitions of the stockflow) unless I'm missing something.

``` julia
julia> macro test_params(a, b, c, block)
         println(a); println(b); println(c); println(block)
       end
@test_params (macro with 1 method)

julia> @test_params 1 2 3 begin
         :test
       end
1
2
3
begin
    #= REPL[5]:2 =#
    :test
end
```

P.S. You may want to enclose stuff like `@stratify` in backticks -- \` -- if you'd like to avoid tagging random users... although those do look like pretty inactive users. 🤷🏻‍♀️

Reply:
It may be me misunderstanding how macros work.  I don't know how to give the stockflows as arguments without using an eval somewhere.  Providing the variables they're bound to just makes them symbols.  I got around this in an earlier draft by checking the Main namespace and grabbing the corresponding value for the symbol, but this seemed like bad practice.

Reply:
This is related to the change in #80 about returning a code block instead of a value. That was a really big mistake on my part to begin with, which I repeated above in my other comment. 

``` julia
julia> macro params2(a, b, c, body)
         return quote
           println($a); println($b); println($c); println($body);
         end
       end
@params2 (macro with 1 method)

julia> @params2 xs ys zs begin
         :test
       end
[1, 2, 3]
[3, 4, 5]
["a", "b", "c"]
test
```

I think this goes back to a while back when I said macros should only rearrange code, not run stuff. By doing it this way, the quoted portion replaces the macro call, then the code that was generated is evaluated. With what I was doing before, the macro call is replaced by a _value_ which is then _evaluated_, and `eval(value)` for any value is just the value.

Alas, I am not a good or experienced Julia programmer.

Reply:
Oh my gosh I didn't know you could do that

Ok should be pretty trivial to add a `@stratify` macro which calls sfstratify.  May want to add another test or two, then...

Should both the macro and the sftratify method be exported, or just the macro?

Reply:
I think just the macro, unless there's a use case for the function for users.

Reply:
I'd need to extract values out of an expression, though.  Don't know if there's a better way...

Reply:
Can a macro take kwargs...?
Might want to keep the function for that

Reply:
> I'd need to extract values out of an expression, though. Don't know if there's a better way...
If I understand you correctly, it may be helpful to look at the code in #80 to get the idea. You can still do stuff like call functions, map over values, assign local variables, etc. in a quote block: https://github.com/AlgebraicJulia/StockFlow.jl/pull/80/files#diff-4e1b9ec77867b43e29dfce6850ad2d9cab1373bef6eed4a38ac4b5c29f94c2a7R147

> Can a macro take kwargs...?
> Might want to keep the function for that

You're right, they cannot. Hmm. I'll have to think about it. It might be worth us discussing this with @Xiaoyan-Li around the final design decisions for the flags/kwargs.

Reply:
Have you confirmed your macros in #80 work?  I'm struggling to get this one working, it keeps saying the variable I'm calling it with is undefined.  I suspect it's a scope issue, since I'm able to get a macro working if I declare it in the same file.  That is, I'm wondering if it's trying to find a variable in the Stratification module, when it only exists in the testing file which calls it.

eg, if I have in Stratification.jl:
```julia
macro stratify(strata, type, aggregate, block)

   return (quote
        $strata
    end)


end
```

and in the testing file I have:

```julia
Weight Model = # insert definition here
@stratify WeightModel l_type ageWeightModel quote ... end
```

I get `  UndefVarError: `WeightModel` not defined`.  However, if I have in one file:

```julia
#!/bin/env julia

macro f(x)
    quote
    cos($x)
    end
end
 begin
x = pi
end

println((@f x))
```

Then this correctly prints -1.0.

I don't understand how you're able to give the block as an argument without it being undefined.

Reply:
Ah, nevermind, I think I got it, I needed an esc.